### ISSUE TITLE
Long torchscript warmup times can be problematic for production serving

### ISSUE THREAD

**Issue by nairbv** [pytorch/pytorch#33354]
Torchscript models take some time to compile on their first query. Initial warmup can be triggered via shadow traffic, but in a large scale elastic cluster of nodes serving many torchscript models, this potentially makes orchestration or user experience more difficult.

In one particular case (densenet 161) we saw warmup times of over 1 minute.

Potential improvements:
* Off-thread compilation (run initial queries unoptimized while compiling in another thread).
* Improve compile times for particular outliers like densenet161

cc @suo , @fbbradheintz 

Related to:
https://github.com/pytorch/pytorch/issues/27610

Example of slow initial query attached, with output:
```
$ python test_torchscript.py
loading model
model loaded
warmup
['n02123045', 'tabby']
first query took 109.52137017250061
starting
['n02123045', 'tabby']
['n02123045', 'tabby']
['n02123045', 'tabby']
['n02123045', 'tabby']
['n02123045', 'tabby']
--- 1.9634242057800293 seconds ---
```

**Comment by harshbafna** [comment]
@nairbv @suo 

I tried this on PyTorch 1.5.1 and 1.6.0. It seems the issue was resolved on PyTorch 1.5.1 and got introduced again in PyTorch 1.6

Used following script to save densenet161 model in script mode:

```
import torch
import torchvision
d161 = torchvision.models.densenet161(pretrained=True)
tsd161 = torch.jit.script(d161)
tsd161.save('tsd161.pt')
```

Then used the following script to run inference on densenet161-scripted model :

```
import torch
import io
import torchvision.transforms as transforms
from PIL import Image
import time


start_time = time.time()


def transform_image(image_bytes):
    my_transforms = transforms.Compose([transforms.Resize(255),
                                        transforms.CenterCrop(224),
                                        transforms.ToTensor(),
                                        transforms.Normalize(
                                            [0.485, 0.456, 0.406],
                                            [0.229, 0.224, 0.225])])
    image = Image.open(io.BytesIO(image_bytes))
    return my_transforms(image).unsqueeze(0)

import json

imagenet_class_index = json.load(open('index_to_name.json'))

def get_prediction(image_bytes):
    tensor = transform_image(image_bytes=image_bytes)
    outputs = model.forward(tensor)
    _, y_hat = outputs.max(1)
    predicted_idx = str(y_hat.item())
    return imagenet_class_index[predicted_idx]


model = torch.jit.load('tsd161.pt')
model.eval()

with open("kitten.jpg", 'rb') as f:
    image_bytes = f.read()
    print(get_prediction(image_bytes=image_bytes))

print("--- %s seconds ---" % (time.time() - start_time))
```


 - Output on torch 1.6 and torchvision 0.7.0

```
ubuntu@ip-172-31-30-88:~/scripted_test$ time python test_torchscript.py 
['n02123045', 'tabby']
--- 170.98425316810608 seconds ---
real 2m51.936s
user 2m52.158s
sys 0m0.448s
```

 - Output on torch 1.5.1 and torchvision 0.6.1

```
ubuntu@ip-172-31-30-88:~/scripted_test$ time python test_torchscript.py 
['n02123045', 'tabby']
--- 2.380345582962036 seconds ---
real 0m3.528s
user 0m3.628s
sys 0m0.614s
```

The inference on scripted models in pytorch 1.5.1 seems superfast compared to pytorch 1.6. 

Am I missing out something?

**Comment by jeremiahschung** [comment]
Hi @nairbv, this issue still persists in the latest build. Is there an ETA on a fix? This will eventually come up more as we have more TorchServe deployments. cc @chauhang

**Comment by eellison** [comment]
Hi, I worked on improving compilation times for the 1.7 release. I was looking at Densenet in particular and got the compilation times for the first run down from 56s -> ~1s. If you're still experiencing bad compilation issues on the latest master could you please comment with a repro ?

**Comment by jeremiahschung** [comment]
@eellison Thanks for the comment! I misspoke when I said the latest build in my comment above. The result before was with the stable release I believe.

I just retested with PT1.7 with the latest commit after building from source. (commit b4ba66a)
Running a slightly modified version of @harshbafna 's script above, the issue has resolved. Initial load time decreased from 181 seconds -> 1.5 on my laptop:

```
import torch
import io
import torchvision.transforms as transforms
from PIL import Image
import time


start_time = time.time()


def transform_image(image_bytes):
    my_transforms = transforms.Compose([transforms.Resize(255),
                                        transforms.CenterCrop(224),
                                        transforms.ToTensor(),
                                        transforms.Normalize(
                                            [0.485, 0.456, 0.406],
                                            [0.229, 0.224, 0.225])])
    image = Image.open(io.BytesIO(image_bytes))
    return my_transforms(image).unsqueeze(0)

import json

imagenet_class_index = json.load(open('index_to_name.json'))

def get_prediction(image_bytes):
    tensor = transform_image(image_bytes=image_bytes)
    outputs = model.forward(tensor)
    _, y_hat = outputs.max(1)
    predicted_idx = str(y_hat.item())
    return imagenet_class_index[predicted_idx]

print("load model")
model = torch.jit.load('tsd161.pt')
model.eval()

with open("kitten.jpg", 'rb') as f:
    image_bytes = f.read()
    print(get_prediction(image_bytes=image_bytes))
print("Run 1:")
print("--- %s seconds ---" % (time.time() - start_time))

start_time = time.time()
with open("kitten.jpg", 'rb') as f:
    image_bytes = f.read()
    print(get_prediction(image_bytes=image_bytes))
print("Run 2:")
print("--- %s seconds ---" % (time.time() - start_time))



start_time = time.time()
with open("kitten.jpg", 'rb') as f:
    image_bytes = f.read()
    print(get_prediction(image_bytes=image_bytes))
print("Run 3:")
print("--- %s seconds ---" % (time.time() - start_time))


start_time = time.time()
with open("kitten.jpg", 'rb') as f:
    image_bytes = f.read()
    print(get_prediction(image_bytes=image_bytes))
print("Run 4:")
print("--- %s seconds ---" % (time.time() - start_time))
```

```
load model
['n02123045', 'tabby']
Run 1:
--- 1.5530858039855957 seconds ---
['n02123045', 'tabby']
Run 2:
--- 0.19954609870910645 seconds ---
['n02123045', 'tabby']
Run 3:
--- 0.17100882530212402 seconds ---
['n02123045', 'tabby']
Run 4:
--- 0.18537425994873047 seconds ---
```

**Comment by eellison** [comment]
@jeremiahschung as of yesterday,  the default on master was the simple executor, not the profiling executor. This has since been updated on master & profiling will be the default for 1.7.

To get a more accurate repro you can either set :
```
self.prev_exec = torch._C._jit_set_profiling_executor(True)
self.prev_profiling = torch._C._jit_set_profiling_mode(True)
```
and rerun the repro, or you can update to the latest master.

**Comment by jeremiahschung** [comment]
@eellison Thanks for the response. I've rerun the tests and verified that both cases are performant. 

FYI - I did note that when the simple executor was run, the first run was ~2.5s with subsequent runs > 0.2s whereas with the profiling executor, the first two runs clocked about 1.5s, 1.0s and it dropped to >0.2s starting with the third run.

Simple executor
```
torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(True)
```
```
Run 1
['n02123045', 'tabby']
--- 2.4346847534179688 seconds ---
['n02123045', 'tabby']
--- 0.17511606216430664 seconds ---
['n02123045', 'tabby']
--- 0.15340709686279297 seconds ---
['n02123045', 'tabby']
--- 0.1473240852355957 seconds ---

Run 2
['n02123045', 'tabby']
--- 2.4973368644714355 seconds ---
['n02123045', 'tabby']
--- 0.17257022857666016 seconds ---
['n02123045', 'tabby']
--- 0.17516398429870605 seconds ---
['n02123045', 'tabby']
--- 0.16097807884216309 seconds ---
```
Profiling executor
```
torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(True)
```
```
Run 1
['n02123045', 'tabby']
--- 1.5280327796936035 seconds ---
['n02123045', 'tabby']
--- 1.0587317943572998 seconds ---
['n02123045', 'tabby']
--- 0.1723308563232422 seconds ---
['n02123045', 'tabby']
--- 0.16199779510498047 seconds ---

Run 2
['n02123045', 'tabby']
--- 1.6480450630187988 seconds ---
['n02123045', 'tabby']
--- 1.0107340812683105 seconds ---
['n02123045', 'tabby']
--- 0.19149112701416016 seconds ---
['n02123045', 'tabby']
--- 0.1593339443206787 seconds ---
```

**Comment by jeremiahschung** [comment]
Confirmed fixed in 1.7 RC.
```
pip install -f https://download.pytorch.org/whl/test/cu101/torch_test.html torch==1.7.0+cu101
```
```
>>> import torch
>>> print(torch.__version__)
1.7.0+cu101
```
```
['n02123045', 'tabby']
--- 0.930553674697876 seconds ---
['n02123045', 'tabby']
--- 0.7171754837036133 seconds ---
```

**Comment by sharvil** [comment]
@eellison can you please point at some of the commits that improved compilation time in the 1.7 release? I'm on 1.8 and am seeing a TorchScript model take >12s when the eager equivalent for the same input runs in ~100ms. Repeated executions of the TorchScript model on the exact same input seems to somewhat arbitrarily choose between a fast path and a slow path and it's unclear to me why that's the case.

**Comment by sharvil** [comment]
Quick update on this issue: the profiling graph executor is recompiling on every second run with the same input data. I haven't had a chance to read through the profiling executor code yet but I assume that it's collecting profiling data in the first run and using that to guide optimizations on the second run.

Switching over to the non-profiling graph optimizer with `torch._C._jit_set_profiling_executor(False)` eliminates the extra recompilations. In my specific case, the profiling optimizer doesn't seem to produce significant performance wins.

I agree with @nairbv that background compilations are the way to go. Repeated blocking compilations on the critical path are a non-starter for deployment.

**Comment by Dango233** [comment]
We have run into a same problem with the Stable Diffusion model using diffusers. The warmup could take ~90s and it's very problematic for production deployment. 

Is there anyway that we can serialize the warmed up model? That would be extremely helpful.

**Comment by domef** [comment]
> Quick update on this issue: the profiling graph executor is recompiling on every second run with the same input data. I haven't had a chance to read through the profiling executor code yet but I assume that it's collecting profiling data in the first run and using that to guide optimizations on the second run.
> 
> Switching over to the non-profiling graph optimizer with `torch._C._jit_set_profiling_executor(False)` eliminates the extra recompilations. In my specific case, the profiling optimizer doesn't seem to produce significant performance wins.
> 
> I agree with @nairbv that background compilations are the way to go. Repeated blocking compilations on the critical path are a non-starter for deployment.

I tried using `torch._C._jit_set_profiling_executor(False)` and it seems to alleviate the warmup problem. I will do some benchmarks to understand if it affects the performance.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
